src/nvim/eval.c

```diff
@@ -10717,55 +10727,62 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv)
     return;
   }
 
-  if (argvars[0].v_type != VAR_STRING
-      || argvars[1].v_type != VAR_STRING
-      || (argvars[2].v_type != VAR_LIST && argvars[2].v_type != VAR_UNKNOWN)) {
+  if (argvars[0].v_type != VAR_LIST
+      || (argvars[1].v_type != VAR_DICT && argvars[1].v_type != VAR_UNKNOWN)) {
     // Wrong argument types
     EMSG(_(e_invarg));
     return;
   }
 
-  argsl = 0;
-  if (argvars[2].v_type == VAR_LIST) {
-    args = argvars[2].vval.v_list;
-    argsl = args->lv_len;
-    // Assert that all list items are strings
-    for (arg = args->lv_first; arg != NULL; arg = arg->li_next) {
-      if (arg->li_tv.v_type != VAR_STRING) {
-        EMSG(_(e_invarg));
-        return;
-      }
+  list_T *args = argvars[0].vval.v_list;
+  // Assert that all list items are strings
+  for (listitem_T *arg = args->lv_first; arg != NULL; arg = arg->li_next) {
+    if (arg->li_tv.v_type != VAR_STRING) {
+      EMSG(_(e_invarg));
+      return;
     }
   }
 
-  if (!os_can_exe(get_tv_string(&argvars[1]), NULL)) {
-    // String is not executable
-    EMSG2(e_jobexe, get_tv_string(&argvars[1]));
+  int argc = args->lv_len;
+  if (!argc) {
+    EMSG(_("Argument vector must have at least one item"));
     return;
   }
 
-  // Allocate extra memory for the argument vector and the NULL pointer
-  argvl = argsl + 2;
-  argv = xmalloc(sizeof(char_u *) * argvl);
-
-  // Copy program name
-  argv[0] = xstrdup((char *)get_tv_string(&argvars[1]));
+  if (!os_can_exe(args->lv_first->li_tv.vval.v_string, NULL)) {
+    // String is not executable
+    EMSG2(e_jobexe, args->lv_first->li_tv.vval.v_string);
+    return;
+  }
 
-  i = 1;
-  // Copy arguments to the vector
-  if (argsl > 0) {
-    for (arg = args->lv_first; arg != NULL; arg = arg->li_next) {
-      argv[i++] = xstrdup((char *)get_tv_string(&arg->li_tv));
+  typval_T user;
+  init_tv(&user);
+  dict_T *job_opts = NULL;
+  ufunc_T *on_stdout = NULL, *on_stderr = NULL, *on_exit = NULL;
+  if (argvars[1].v_type == VAR_DICT) {
+    job_opts = argvars[1].vval.v_dict;
+    common_job_callbacks(job_opts, &on_stdout, &on_stderr, &on_exit, &user);
+    if (did_emsg) {
+      return;
     }
   }
 
-  // The last item of argv must be NULL
-  argv[i] = NULL;
-  JobOptions opts = common_job_options(argv, (char *)argvars[0].vval.v_string);
+  // Build the argument vector
+  int i = 0;
+  char **argv = xcalloc(argc + 1, sizeof(char *));
+  for (listitem_T *arg = args->lv_first; arg != NULL; arg = arg->li_next) {
+    argv[i++] = xstrdup((char *)arg->li_tv.vval.v_string);
+  }
```

Comment:
`v_string` may be `NULL` in some circumstances to represent an empty string. `get_tv_string()` would return an empty string in this case, and also allow the implicit conversion of integers to strings. Same comment for the `os_can_exe()` call above.

ref: #1471 for an instance where this may segfault.